google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDMLTest.java

```diff
@@ -363,6 +414,98 @@ public void standardDMLAndMutations() {
     assertThat(rowCount).isEqualTo(2);
   }
 
+  @Test
+  public void testUpsertDml() {
+    executeUpdate(DML_COUNT_WITH_UPSERT, insertOrIgnoreDml(dialect.dialect));
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(1);
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(5);
+    executeUpdate(DML_COUNT_WITH_UPSERT, insertOrUpdateDml(dialect.dialect));
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(10);
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(50);
+    executeUpdate(DML_COUNT_WITH_UPSERT, deleteDml());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
+        .isNull();
+  }
+
```

Comment:
We can add 2 more tests as below. They cover cases like:
1. Insert or ignore ignores existing rows.
2. Insert or update inserts (not updates) in case of new rows. 

 @Test
  public void testInsertOrUpdateOnNewRow() {
    if (dialect.dialect == Dialect.GOOGLE_STANDARD_SQL) {
      executeUpdate(DML_COUNT_WITH_UPSERT, insertOrUpdateDml());
    } else {
      executeUpdate(DML_COUNT_WITH_UPSERT, insertOrUpdateDmlPostgresql());
    }
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
                .getLong(0))
        .isEqualTo(10);
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
                .getLong(0))
        .isEqualTo(50);
    // Delete the rows
    executeUpdate(DML_COUNT_WITH_UPSERT, deleteDml());
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
        .isNull();
  }

  @Test
  public void testInsertOrIgnoreDmlOnExistingRow() {
    // First insert few rows.
    executeUpdate(DML_COUNT, insertDml());
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
                .getLong(0))
        .isEqualTo(1);

    // Existing rows are ignored, only 1 new row is inserted.
    if (dialect.dialect == Dialect.GOOGLE_STANDARD_SQL) {
      executeUpdate(1, insertOrIgnoreDml());
    } else {
      executeUpdate(1, insertOrIgnoreDmlPostgresql());
    }
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
                .getLong(0))
        .isEqualTo(1);
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
                .getLong(0))
        .isEqualTo(5);
    // Delete the rows
    executeUpdate(DML_COUNT_WITH_UPSERT, deleteDml());
    assertThat(
            getClient(dialect.dialect)
                .singleUse(TimestampBound.strong())
                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
        .isNull();
  }